lib/fetch/formdata.js

```diff
@@ -157,8 +157,11 @@ class FormData {
     // replace the first such entry with entry and remove the others.
     const idx = this[kState].findIndex((entry) => entry.name === name)
     if (idx !== -1) {
-      this[kState][idx] = entry
-      this[kState] = this[kState].filter((entry) => entry.name !== name)
+      this[kState] = [].concat(
+        this[kState].slice(0, idx).filter((entry) => entry.name !== name),
+        [entry],
+        this[kState].slice(idx).filter((entry) => entry.name !== name),
+      )
```

Comment:
```suggestion
      this[kState][idx] = entry
```

Isn't this sufficient?

Reply:
i.e just remove `this[kState] = this[kState].filter((entry) => entry.name !== name)`

Reply:
> i.e just remove `this[kState] = this[kState].filter((entry) => entry.name !== name)`

But this wil not remove all other entries which names are also `<name>`, as described in the comment above.

Reply:
`this[kState]` could be something like
```javascript
[
    { name: "k1", value: "v1" },
    { name: "k2", value: "v2" },
    { name: "k1", value: "v3" },  // another entry which name is `k1`
]
```
Probably it's why `FormData.getAll()` exists?

Reply:
And also, perhaps `form.append` should insert entries at the beginning of `this[kState]` if they are dup key entries. Otherwise, this code fragment will be a little weird.
```javascript
const form = new FormData()
form.set('k1', 'v1')
form.append('k1', 'v2')
form.get('k1') // we might want it to be 'v2', but get 'v1', weird.
```
How do you think of it?

Reply:
The spec says 'replace the first such entry with entry and remove the others.'

Reply:
> The spec says 'replace the first such entry with entry and remove the others.'

Indeed. That's exactly what this PR does.

Perhaps I didn't make myself clear before. Let me be more specific.

Consider this following code snippet
```javascript
const form = new FormData()
form.append('k1', 'v1')
form.append('k2', 'v2')
form.append('k1', 'v3')
form.set('k1', 'v4')
```
The spec in the method `FormData.set` says (according to my understanding) it will replace the first matched entry with the new created entry, and remove all other entries sharing the same name.

After running the code snippet above, we expect `this[kState]` to be

```javascript
[
    { name: 'k1', value: 'v4' },
    { name: 'k2', value: 'v2' },
]
```

but currently we get
```javascript
[
    { name: 'k1', value: 'v4' },
]
```
This is because line 161 empties the entire `this[kState]` array.
https://github.com/nodejs/undici/blob/1c2259cf7a89ed427cf915719ee301ef7751b4eb/lib/fetch/formdata.js#L158-L166

if we just delete `this[kState] = this[kState].filter((entry) => entry.name !== name)` as you suggested previously, we get
```javascript
[
    { name: 'k1', value: 'v4' },
    { name: 'k2', value: 'v2' },
    { name: 'k1', value: 'v3' },
]
```
according to the spec described, we need to remove `{ name: 'k1', value: 'v3' }`, but we didn't.

Reply:
You are right.